torch/csrc/autograd/graph_task.h

```diff
@@ -0,0 +1,191 @@
+  // mark the graph task as completed and trigger post processing
+  void mark_as_completed_and_run_post_processing();
+
+  // Set an appropriate exception on this graph_task which was encountered while
+  // running the provided function.
+  void set_exception(std::exception_ptr eptr, const std::shared_ptr<Node>& fn);
+
+  // Set an appropriate exception on this graph_task which was encountered while
+  // running the provided function. But doesn't signal completion on
+  // 'future_result_' right away. The user needs to explicitly mark
+  // 'future_result_' completed with an appropriate exception.
+  void set_exception_without_signal(const std::shared_ptr<Node>& fn);
+
+  // Whether or not to stop execution for this GraphTask when an error is
+  // encountered. When set to true, this would cause Engine::execute() to throw
+  // an exception as soon as the autograd engine receives an exception.
+  bool exit_on_error_;
+
+  // CPU threads are dedicated to processing CPU work for the backward they
+  // invoked. So any given graph task maintains its own cpu_ready_queue_ where
+  // you should send work for it to be done. We memoize the cpu_ready_queue_ per
+  // GraphTask so that we know which ready queue we should push to if we are on
+  // device thread (i.e. GPU) and but next NodeTask should be run on CPU.
+  std::shared_ptr<ReadyQueue> cpu_ready_queue_;
+
+  // Future representing the completion of the graph task. Notified when all
+  // tasks are done.
+  c10::intrusive_ptr<at::ivalue::Future> future_result_;
+
+  // Final callbacks installed during execution of this GraphTask
+  std::vector<std::function<void()>> final_callbacks_;
+  // To protect reads and writes to final_callbacks_. Intentionally no reusing
+  // mutex_ as the two are protecting different data structures.
+  std::mutex final_callbacks_lock_;
+
+  utils::DelayWarningHandler warning_handler_;
+
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+  GraphTask(
+      bool keep_graph,
+      bool grad_mode,
+      int reentrant_depth,
+      std::shared_ptr<ReadyQueue> cpu_ready_queue,
+      bool exit_on_error = false)
+      : keep_graph_(keep_graph),
+        owner_(NO_DEVICE),
+        reentrant_depth_(reentrant_depth),
+        exit_on_error_(exit_on_error),
+        cpu_ready_queue_(std::move(cpu_ready_queue)),
+        future_result_(c10::make_intrusive<at::ivalue::Future>(
+            c10::ListType::create(c10::TensorType::get()))) {
+    thread_locals_.set_grad_mode(grad_mode);
+  }
+
+ private:
+  // run GraphTask post processing
+  void exec_post_processing();
+};
+
+// The guard that sets and restores current_graph_task.
+class GraphTaskGuard {
+ public:
+  explicit GraphTaskGuard(std::shared_ptr<GraphTask> graph_task);
+  ~GraphTaskGuard();
+
+  void restore_current_graph_task();
+
+ private:
+  std::shared_ptr<GraphTask> last_graph_task_;
+};
+
+TORCH_API const std::unordered_map<Node*, GraphTask::ExecInfo>*
+get_current_graph_task_exec_info();
```

Comment:
`TORCH_API` is needed for functorch, otherwise we got `undefined symbol` error

Reply:
I don't actually understand this too well - afaik TORCH_API is used for windows builds? cc @zou3519

Reply:
I also changed the `functorch/functorch/csrc/CustomFunction.cpp` ([diff](https://github.com/pytorch/pytorch/pull/82544/files#diff-e64a5729778d2216339c0eb320a42225e1213976a485bd095bab1ec6886b6c8fL195)), and functorch also calls `task_should_compute_output`. Because functorch build it's own `_C.so`, we have to export the symbol `get_current_graph_task_exec_info ` from `libtorch*.so`, otherwise we got `undefined symbol` error when calling `task_should_compute_output` from functorch.

Reply:
Hmm interesting.  My understanding was that symbols are exported by default on Linux, so this failure would be windows specific - 

https://github.com/pytorch/pytorch/blob/0985188c8fa2ffe8cba7c8a1bd32ec6d045940d3/CONTRIBUTING.md?plain=1#L934-L950

but apparently it is not? https://github.com/pytorch/pytorch/runs/7683263397 is failing on Linux

Reply:
I think what you're referring to is visibility which is not the same as exporting even if unused.

Reply:
Yeah it is a visibility thing, found the cmake config
https://github.com/pytorch/pytorch/blob/f6c275f55ddfd22f8c0558efad3068b0e91f1560/cmake/public/utils.cmake#L500-L510

Reply:
Thanks for correcting me! Does that mean I should not use `TORCH_API` here?

Reply:
You should as we need to export the function for people outside of the lib.